tools/offline_log_viewer/reader.py

```diff
@@ -127,32 +127,50 @@ def read_map(self, k_reader, v_reader):
             ret[key] = val
         return ret
 
-    def read_envelope(self, type_read=None, max_version=0):
+    def read_envelope(self, type_read=None, max_compat_version=0):
         header = self.read_bytes(SERDE_ENVELOPE_SIZE)
         envelope = SerdeEnvelope(*struct.unpack(SERDE_ENVELOPE_FORMAT, header))
-        return self.read_envelope_inner(envelope, type_read, max_version)
+        return self.read_envelope_inner(envelope, type_read,
+                                        max_compat_version)
 
-    def read_checksum_envelope(self, type_read=None, max_version=0):
+    def read_checksum_envelope(self, type_read=None, max_compat_version=0):
         header = self.read_bytes(SERDE_CHECKSUM_ENVELOPE_SIZE)
         envelope = SerdeChecksumEnvelope(
             *struct.unpack(SERDE_CHECKSUM_ENVELOPE_FORMAT, header))
-        return self.read_envelope_inner(envelope, type_read, max_version)
+        return self.read_envelope_inner(envelope, type_read,
+                                        max_compat_version)
+
+    def read_envelope_inner(self, envelope, type_read, max_compat_version):
+        bytes_left_in_envelope = envelope.size
+        envelope_start_pos = self.stream.tell()
 
-    def read_envelope_inner(self, envelope, type_read, max_version):
         if type_read is not None:
-            if envelope.version <= max_version:
+            if envelope.version > max_compat_version:
+                logger.warning(
+                    f"Reading {type_read=} envelope with version {envelope.version}, but max compat version is {max_compat_version}."
+                    " The output is likely to be incomplete.")
+
+            if envelope.compat_version <= max_compat_version:
                 v = type_read(self, envelope.version)
+
+                # Skip the rest of the envelope.
+                bytes_left_in_envelope -= self.stream.tell(
+                ) - envelope_start_pos
+                self.stream.read(bytes_left_in_envelope)
```

Comment:
Do we need to cover this case: https://github.com/redpanda-data/redpanda/blob/3f944f5311ad1f342c6c65922ce44dbcb838f8f1/src/v/serde/read_header.h#L75

Like if the `type_read` implies a compat version that exceeds the encoded version of the binary blob we're reading, should we reject that as well?